### PR TITLE
feat: allows forcing/skipping calendar cache serving

### PR DESCRIPTION
**Pull request by zomars**
Folow up to #11928

## What does this PR do?

This pull request introduces several enhancements and refactors to the calendar service, focusing on caching and availability fetching mechanisms. The key changes include the introduction of a new `shouldServeCache` parameter, refactoring of cache-related functions, and updates to various calendar service methods to incorporate these changes.

Enhancements to caching and availability fetching:

* [`packages/app-store/googlecalendar/lib/CalendarService.ts`](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L513-R529): Introduced the `shouldServeCache` parameter in multiple methods, refactored `getCacheOrFetchAvailability` to use a new helper method `getFreeBusyResult`, and updated cache handling in `unwatchCalendar`. [[1]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L513-R529) [[2]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L566-R581) [[3]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L586-R604) [[4]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L675-R689) [[5]](diffhunk://#diff-6bfbaf63c841dc9343dba0c7487e30a546ffe2a63c9d9418379d0dbb88644d27L710-R729)

* [`packages/core/CalendarManager.ts`](diffhunk://#diff-1edf77640cddbeb991fcd8211c16508c055ff086f3c04f58a4e951f1e6710b14L139-L183): Removed the `getCachedResults` function and added the `shouldServeCache` parameter to `getBusyCalendarTimes` and related functions to improve cache handling. [[1]](diffhunk://#diff-1edf77640cddbeb991fcd8211c16508c055ff086f3c04f58a4e951f1e6710b14L139-L183) [[2]](diffhunk://#diff-1edf77640cddbeb991fcd8211c16508c055ff086f3c04f58a4e951f1e6710b14L206-R161) [[3]](diffhunk://#diff-1edf77640cddbeb991fcd8211c16508c055ff086f3c04f58a4e951f1e6710b14L215-R176)

* [`packages/core/getBusyTimes.ts`](diffhunk://#diff-d3d426265221c77cb2fa5ccde7a3cfcf61db38f62011c3653e97c0aa79096afeR47): Added the `shouldServeCache` parameter to the `getBusyTimes` function to control cache usage. [[1]](diffhunk://#diff-d3d426265221c77cb2fa5ccde7a3cfcf61db38f62011c3653e97c0aa79096afeR47) [[2]](diffhunk://#diff-d3d426265221c77cb2fa5ccde7a3cfcf61db38f62011c3653e97c0aa79096afeR64) [[3]](diffhunk://#diff-d3d426265221c77cb2fa5ccde7a3cfcf61db38f62011c3653e97c0aa79096afeL246-R249)

* [`packages/core/getCalendarsEvents.ts`](diffhunk://#diff-b15e52ad5aa461f1c933ef3a852dcfaf5278fa7798855e238f01cba46207adf8L14-R15): Updated the `getCalendarsEvents` function to include the `shouldServeCache` parameter for better cache management. [[1]](diffhunk://#diff-b15e52ad5aa461f1c933ef3a852dcfaf5278fa7798855e238f01cba46207adf8L14-R15) [[2]](diffhunk://#diff-b15e52ad5aa461f1c933ef3a852dcfaf5278fa7798855e238f01cba46207adf8L48-R54)

* [`packages/features/calendar-cache/lib/datesForCache.ts`](diffhunk://#diff-394e8011261a8266dcd61e91e2401d8c31708d7a2b8448ca87e2ef9fd5ee8f19R1-R16): Added new utility functions `getTimeMin` and `getTimeMax` to handle date range expansions for caching purposes.

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### Scenario 1

- Have a team with a TeamFeature entry for both "calendar-cache" and "calendar-cache-serve"
- Test out a booking normally
- Test out a booking with `?cal.cache=false` parameter. It should skip cache altogether.

### Scenario 2

- Have a team with a TeamFeature entry for both "calendar-cache"only.
- Test out a booking normally
- Test out a booking with `?cal.cache=true` parameter. It should force cache serving.

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

**Review by zomars — COMMENTED** [review]
Ready for review

File: packages/core/getBusyTimes.ts [hunk 1]
```diff
@@ -44,6 +44,7 @@ export async function getBusyTimes(params: {
       })[]
     | null;
   bypassBusyCalendarTimes: boolean;
+  shouldServeCache?: boolean;
```
Comment (keithwillcode) [inline packages/core/getBusyTimes.ts:47]: Why is this nullable? Seems like we should either just pass true or false.

File: packages/features/calendar-cache/lib/getShouldServeCache.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { FeaturesRepository } from "@calcom/features/flags/features.repository";
+
+export async function getShouldServeCache(shouldServeCache?: boolean | undefined, teamId?: number) {
+  if (typeof shouldServeCache === "boolean") return shouldServeCache;
+  if (!teamId) return undefined;
```
Comment (keithwillcode) [inline packages/features/calendar-cache/lib/getShouldServeCache.ts:5]: Why not `false` here?

File: packages/prisma/migrations/20241216000000_add_calendar_cache_serve/migration.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO
+  "Feature" (slug, enabled, description, "type")
+VALUES
+  (
+    'calendar-cache-serve',
+    false,
+    'Wether to serve calendar cache by default or not per team/user basis.',
```
Comment (keithwillcode) [inline packages/prisma/migrations/20241216000000_add_calendar_cache_serve/migration.sql:7]:
```suggestion
    'Whether to serve calendar cache by default or not on a team/user basis.',
```

File: packages/prisma/zod-utils.ts [hunk 4]
```diff
@@ -265,6 +265,8 @@ export const bookingCreateBodySchema = z.object({
    * Used to identify if the booking is a dry run.
    */
   _isDryRun: z.boolean().optional(),
+  /** Wether to override the cache */
```
Comment (keithwillcode) [inline packages/prisma/zod-utils.ts:268]:
```suggestion
  /** Whether to override the cache */
```

File: packages/core/getCalendarsEvents.ts [hunk 5]
```diff
@@ -11,7 +11,8 @@ const getCalendarsEvents = async (
   withCredentials: CredentialPayload[],
   dateFrom: string,
   dateTo: string,
-  selectedCalendars: SelectedCalendar[]
+  selectedCalendars: SelectedCalendar[],
+  shouldServeCache?: boolean
```
[the remark this answers is not in the document]
Comment (zomars) [reply packages/core/getCalendarsEvents.ts:15]: https://github.com/calcom/cal.com/pull/18224#discussion_r1887900863

[the remark this answers is not in the document]
Comment (zomars) [reply packages/core/getCalendarsEvents.ts:15]: Undefined will fallback to the DB default. If we force a boolean value the fallback will never execute.

File: packages/features/calendar-cache/lib/getShouldServeCache.ts (hunk 2 above)
Comment (zomars) [reply packages/features/calendar-cache/lib/getShouldServeCache.ts:5]: Undefined means there's no preference. So the regular default behavior will apply.

**Review by tarunkumarmetaforms — APPROVED** [review]

File: packages/features/flags/features.repository.ts [hunk 6]
```diff
@@ -65,4 +65,15 @@ export class FeaturesRepository implements IFeaturesRepository {
       throw err;
     }
   }
+  async checkIfTeamHasFeature(teamId: number, featureId: keyof AppFlags) {
+    try {
+      const teamFeature = await db.teamFeatures.findUnique({
```
Comment (emrysal) [inline packages/features/flags/features.repository.ts:70]: Did you mean?

```suggestion
      const teamFeature = await db.teamFeatures.findUniqueOrThrow({
```

Comment (zomars) [reply packages/features/flags/features.repository.ts:70]: Not throwing since we're expecting a boolean return value

**Review by emrysal — APPROVED** [review]
Happy with this, let's get it merged.

**Pull request merged**